crates/wysiwyg/src/format_nodes.rs

```diff
@@ -0,0 +1,405 @@
+                        indexes_to_remove.push(n.handle().index_in_parent());
+                    }
+                }
+            }
+            for i in indexes_to_remove {
+                parent.remove_child(i);
+            }
+        }
+    }
+
+    fn merge_formatting_nodes(&mut self, parent_handle: DomHandle) {
+        if let DomNode::Container(parent) =
+            self.state.dom.lookup_node_mut(parent_handle.clone())
+        {
+            let children = parent.children();
+            for i in (0..children.len() - 1).rev() {
+                parent.merge_children(i + 1, i);
+            }
+        } else {
+            panic!("Parent node must be a Container.");
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::tests::testutils_composer_model::cm;
+
+    use super::*;
+
+    fn find_range<S: UnicodeString>(model: &ComposerModel<S>) -> Range {
+        let (start, end) = model.get_selection();
+        model.state.dom.find_range(start.into(), end.into())
+    }
+
+    fn find_multiple_range<S: UnicodeString>(
+        model: &ComposerModel<S>,
+    ) -> MultipleNodesRange {
+        let range = find_range(&model);
+        if let Range::MultipleNodes(r) = range {
+            r
+        } else {
+            panic!("Should have been a multiple range node, {:?}", range);
+        }
+    }
+
+    #[test]
+    fn selection_type_extend_if_different_type() {
+        let model = cm("{hello <b>wor}|ld</b>");
+        let range = find_multiple_range(&model);
+        let selection_type = model.check_format_selection_type(
+            &range.locations,
+            &InlineFormatType::Italic,
+        );
+        assert_eq!(selection_type, FormatSelectionType::Extend);
+    }
+
+    #[test]
+    fn selection_type_extending_start() {
+        let model = cm("hell{o <b>w}|orld</b>");
+        let range = find_multiple_range(&model);
+        let selection_type = model.check_format_selection_type(
+            &range.locations,
+            &InlineFormatType::Bold,
+        );
+        assert_eq!(selection_type, FormatSelectionType::Extend);
+    }
+
+    #[test]
+    fn selection_type_extending_end() {
+        let model = cm("<b>hell{o </b>wor}|ld");
+        let range = find_multiple_range(&model);
+        let selection_type = model.check_format_selection_type(
+            &range.locations,
+            &InlineFormatType::Bold,
+        );
+        assert_eq!(selection_type, FormatSelectionType::Extend);
+    }
+
+    #[test]
+    fn selection_type_extending_middle() {
+        let model = cm("<b>h{el</b>lo <b>wor}|ld</b>");
+        let range = find_multiple_range(&model);
+        let selection_type = model.check_format_selection_type(
+            &range.locations,
+            &InlineFormatType::Bold,
+        );
+        assert_eq!(selection_type, FormatSelectionType::Extend);
+    }
+
+    #[test]
+    fn selection_type_remove() {
+        let model = cm("<b>hel{lo </b><b>wor}|ld</b>");
```

Comment:
This example is weird because there are 2 bold nodes next to each other

Reply:
(also a few more similar ones below)

Reply:
True. I'll wrap it into another format type so we can get a `MultipleNodeRange` to verify `check_format_selection_type`.